File: examples/go/sphincs-wallet/SPHINCSWallet_test.go
package contract

import (
	"testing"

	runar "github.com/icellan/runar/packages/runar-go"
)

// deterministic seed for reproducible tests
var testSeed = func() []byte {
	s := make([]byte, 3*16) // 3*n for SLH-DSA-128s (n=16)
	for i := range s {
		s[i] = byte(i)
	}
	return s
}()

var testKP = runar.SLHKeygen(runar.SLH_SHA2_128s, testSeed)

func setupSPHINCSKeys() (ecdsaPubKey runar.PubKey, ecdsaPubKeyHash runar.Addr, slhdsaPubKeyHash runar.ByteString) {
	ecdsaPubKey = runar.MockPubKey()
	ecdsaPubKeyHash = runar.Hash160(ecdsaPubKey)
	slhdsaPubKeyHash = runar.Hash160(runar.ByteString(testKP.PK))
	return
}

func TestSPHINCSWallet_Spend(t *testing.T) {
	ecdsaPubKey, ecdsaPubKeyHash, slhdsaPubKeyHash := setupSPHINCSKeys()

	c := &SPHINCSWallet{
		EcdsaPubKeyHash:  ecdsaPubKeyHash,
		SlhdsaPubKeyHash: slhdsaPubKeyHash,
	}

	// Mock ECDSA signature (CheckSig is mocked to true)
	ecdsaSig := runar.MockSig()

	// SLH-DSA-sign the ECDSA signature bytes
	slhdsaSig := runar.SLHSign(runar.SLH_SHA2_128s, []byte(ecdsaSig), testKP.SK)

	c.Spend(runar.ByteString(slhdsaSig), runar.ByteString(testKP.PK), ecdsaSig, ecdsaPubKey)
}

func TestSPHINCSWallet_Spend_MultipleMessages(t *testing.T) {
	// SLH-DSA is stateless — same keypair can sign many messages
	ecdsaPubKey, ecdsaPubKeyHash, slhdsaPubKeyHash := setupSPHINCSKeys()

	c := &SPHINCSWallet{
		EcdsaPubKeyHash:  ecdsaPubKeyHash,
		SlhdsaPubKeyHash: slhdsaPubKeyHash,
	}

	// First spend
	ecdsaSig1 := runar.Sig(make([]byte, 72))
	ecdsaSig1 = runar.Sig(append([]byte{0x30, 0x01}, make([]byte, 70)...))
	slhdsaSig1 := runar.SLHSign(runar.SLH_SHA2_128s, []byte(ecdsaSig1), testKP.SK)
	c.Spend(runar.ByteString(slhdsaSig1), runar.ByteString(testKP.PK), ecdsaSig1, ecdsaPubKey)

	// Second spend (same SLH-DSA keypair, different ECDSA sig)
	ecdsaSig2 := runar.Sig(append([]byte{0x30, 0x02}, make([]byte, 70)...))
	slhdsaSig2 := runar.SLHSign(runar.SLH_SHA2_128s, []byte(ecdsaSig2), testKP.SK)
	c.Spend(runar.ByteString(slhdsaSig2), runar.ByteString(testKP.PK), ecdsaSig2, ecdsaPubKey)
}

func TestSPHINCSWallet_Spend_TamperedSLHDSA(t *testing.T) {
	ecdsaPubKey, ecdsaPubKeyHash, slhdsaPubKeyHash := setupSPHINCSKeys()

	c := &SPHINCSWallet{
		EcdsaPubKeyHash:  ecdsaPubKeyHash,
		SlhdsaPubKeyHash: slhdsaPubKeyHash,
	}

	ecdsaSig := runar.MockSig()
	slhdsaSig := runar.SLHSign(runar.SLH_SHA2_128s, []byte(ecdsaSig), testKP.SK)
	slhdsaSig[0] ^= 0xff // tamper

	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected spend to fail with tampered SLH-DSA signature")
		}
	}()
	c.Spend(runar.ByteString(slhdsaSig), runar.ByteString(testKP.PK), ecdsaSig, ecdsaPubKey)
}

func TestSPHINCSWallet_Spend_WrongECDSASig(t *testing.T) {
	ecdsaPubKey, ecdsaPubKeyHash, slhdsaPubKeyHash := setupSPHINCSKeys()

	c := &SPHINCSWallet{
		EcdsaPubKeyHash:  ecdsaPubKeyHash,
		SlhdsaPubKeyHash: slhdsaPubKeyHash,
	}

	// Sign one ECDSA sig with SLH-DSA, but provide a different ECDSA sig
	ecdsaSig1 := runar.MockSig()
	slhdsaSig := runar.SLHSign(runar.SLH_SHA2_128s, []byte(ecdsaSig1), testKP.SK)

	ecdsaSig2 := runar.Sig([]byte{0x30, 0xFF})

	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected spend to fail when SLH-DSA signed wrong ECDSA sig")
		}
	}()
	c.Spend(runar.ByteString(slhdsaSig), runar.ByteString(testKP.PK), ecdsaSig2, ecdsaPubKey)
}

func TestSPHINCSWallet_Spend_WrongECDSAPubKeyHash(t *testing.T) {
	_, ecdsaPubKeyHash, slhdsaPubKeyHash := setupSPHINCSKeys()

	c := &SPHINCSWallet{
		EcdsaPubKeyHash:  ecdsaPubKeyHash,
		SlhdsaPubKeyHash: slhdsaPubKeyHash,
	}

	// Different ECDSA pubkey whose hash160 won't match
	wrongBytes := make([]byte, 33)
	wrongBytes[0] = 0x03
	for i := 1; i < 33; i++ {
		wrongBytes[i] = 0xff
	}
	wrongECDSAPubKey := runar.PubKey(wrongBytes)

	ecdsaSig := runar.MockSig()
	slhdsaSig := runar.SLHSign(runar.SLH_SHA2_128s, []byte(ecdsaSig), testKP.SK)

	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected spend to fail with wrong ECDSA public key hash")
		}
	}()
	c.Spend(runar.ByteString(slhdsaSig), runar.ByteString(testKP.PK), ecdsaSig, wrongECDSAPubKey)
}

func TestSPHINCSWallet_Spend_WrongSLHDSAPubKeyHash(t *testing.T) {
	ecdsaPubKey, ecdsaPubKeyHash, slhdsaPubKeyHash := setupSPHINCSKeys()

	c := &SPHINCSWallet{
		EcdsaPubKeyHash:  ecdsaPubKeyHash,
		SlhdsaPubKeyHash: slhdsaPubKeyHash,
	}

	// Different SLH-DSA keypair whose hash160 won't match
	wrongSeed := make([]byte, 3*16)
	for i := range wrongSeed {
		wrongSeed[i] = 0xff
	}
	wrongKP := runar.SLHKeygen(runar.SLH_SHA2_128s, wrongSeed)

	ecdsaSig := runar.MockSig()
	slhdsaSig := runar.SLHSign(runar.SLH_SHA2_128s, []byte(ecdsaSig), wrongKP.SK)

	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected spend to fail with wrong SLH-DSA public key hash")
		}
	}()
	c.Spend(runar.ByteString(slhdsaSig), runar.ByteString(wrongKP.PK), ecdsaSig, ecdsaPubKey)
}

func TestSPHINCSWallet_Compile(t *testing.T) {
	if err := runar.CompileCheck("SPHINCSWallet.runar.go"); err != nil {
		t.Fatalf("Rúnar compile check failed: %v", err)
	}
}

File: examples/rust/sphincs-wallet/SPHINCSWallet_test.rs
#[path = "SPHINCSWallet.runar.rs"]
mod contract;

use contract::*;
use runar::prelude::{hash160, mock_pub_key, mock_sig, slh_keygen, slh_sign, SLH_SHA2_128S};

fn setup_keys() -> (Vec<u8>, Vec<u8>, runar::prelude::SlhKeyPair, Vec<u8>) {
    let ecdsa_pub_key = mock_pub_key();
    let ecdsa_pub_key_hash = hash160(&ecdsa_pub_key);

    let params = &SLH_SHA2_128S;
    let seed = vec![0x42u8; 3 * params.n];
    let kp = slh_keygen(params, Some(&seed));
    let slhdsa_pub_key_hash = hash160(&kp.pk);

    (ecdsa_pub_key, ecdsa_pub_key_hash, kp, slhdsa_pub_key_hash)
}

#[test]
fn test_spend() {
    let (ecdsa_pub_key, ecdsa_pub_key_hash, kp, slhdsa_pub_key_hash) = setup_keys();

    let c = SPHINCSWallet {
        ecdsa_pub_key_hash,
        slhdsa_pub_key_hash,
    };

    // Mock ECDSA signature (check_sig is mocked to true)
    let ecdsa_sig = mock_sig();

    // SLH-DSA-sign the ECDSA signature bytes
    let slhdsa_sig = slh_sign(&SLH_SHA2_128S, &ecdsa_sig, &kp.sk);

    c.spend(&slhdsa_sig, &kp.pk, &ecdsa_sig, &ecdsa_pub_key);
}

#[test]
fn test_spend_multiple_messages() {
    // SLH-DSA is stateless — same keypair can sign many messages
    let (ecdsa_pub_key, ecdsa_pub_key_hash, kp, slhdsa_pub_key_hash) = setup_keys();

    let c = SPHINCSWallet {
        ecdsa_pub_key_hash,
        slhdsa_pub_key_hash,
    };

    let ecdsa_sig1 = vec![0x30, 0x01];
    let slhdsa_sig1 = slh_sign(&SLH_SHA2_128S, &ecdsa_sig1, &kp.sk);
    c.spend(&slhdsa_sig1, &kp.pk, &ecdsa_sig1, &ecdsa_pub_key);

    let ecdsa_sig2 = vec![0x30, 0x02];
    let slhdsa_sig2 = slh_sign(&SLH_SHA2_128S, &ecdsa_sig2, &kp.sk);
    c.spend(&slhdsa_sig2, &kp.pk, &ecdsa_sig2, &ecdsa_pub_key);
}

#[test]
fn test_spend_tampered_slhdsa() {
    let (ecdsa_pub_key, ecdsa_pub_key_hash, kp, slhdsa_pub_key_hash) = setup_keys();

    let c = SPHINCSWallet {
        ecdsa_pub_key_hash,
        slhdsa_pub_key_hash,
    };

    let ecdsa_sig = mock_sig();
    let mut slhdsa_sig = slh_sign(&SLH_SHA2_128S, &ecdsa_sig, &kp.sk);
    slhdsa_sig[0] ^= 0xff; // tamper

    let result = std::panic::catch_unwind(|| c.spend(&slhdsa_sig, &kp.pk, &ecdsa_sig, &ecdsa_pub_key));
    assert!(result.is_err(), "expected spend to fail with tampered SLH-DSA signature");
}

#[test]
fn test_spend_wrong_ecdsa_sig() {
    let (ecdsa_pub_key, ecdsa_pub_key_hash, kp, slhdsa_pub_key_hash) = setup_keys();

    let c = SPHINCSWallet {
        ecdsa_pub_key_hash,
        slhdsa_pub_key_hash,
    };

    // Sign one ECDSA sig with SLH-DSA, but provide different ECDSA sig to contract
    let ecdsa_sig1 = mock_sig();
    let slhdsa_sig = slh_sign(&SLH_SHA2_128S, &ecdsa_sig1, &kp.sk);

    let ecdsa_sig2 = vec![0x30, 0xFF];

    let result = std::panic::catch_unwind(|| c.spend(&slhdsa_sig, &kp.pk, &ecdsa_sig2, &ecdsa_pub_key));
    assert!(result.is_err(), "expected spend to fail when SLH-DSA signed wrong ECDSA sig");
}

#[test]
fn test_spend_wrong_ecdsa_pub_key_hash() {
    let (_, ecdsa_pub_key_hash, kp, slhdsa_pub_key_hash) = setup_keys();

    let c = SPHINCSWallet {
        ecdsa_pub_key_hash,
        slhdsa_pub_key_hash,
    };

    // Different ECDSA pubkey whose hash160 won't match
    let wrong_ecdsa_pub_key = {
        let mut k = vec![0x03u8];
        k.extend_from_slice(&[0xffu8; 32]);
        k
    };

    let ecdsa_sig = mock_sig();
    let slhdsa_sig = slh_sign(&SLH_SHA2_128S, &ecdsa_sig, &kp.sk);

    let result = std::panic::catch_unwind(|| c.spend(&slhdsa_sig, &kp.pk, &ecdsa_sig, &wrong_ecdsa_pub_key));
    assert!(result.is_err(), "expected spend to fail with wrong ECDSA public key hash");
}

#[test]
fn test_spend_wrong_slhdsa_pub_key_hash() {
    let (ecdsa_pub_key, ecdsa_pub_key_hash, _, slhdsa_pub_key_hash) = setup_keys();

    let c = SPHINCSWallet {
        ecdsa_pub_key_hash,
        slhdsa_pub_key_hash,
    };

    // Different SLH-DSA keypair whose hash160 won't match
    let wrong_seed = vec![0xffu8; 3 * SLH_SHA2_128S.n];
    let wrong_kp = slh_keygen(&SLH_SHA2_128S, Some(&wrong_seed));

    let ecdsa_sig = mock_sig();
    let slhdsa_sig = slh_sign(&SLH_SHA2_128S, &ecdsa_sig, &wrong_kp.sk);

    let result = std::panic::catch_unwind(|| c.spend(&slhdsa_sig, &wrong_kp.pk, &ecdsa_sig, &ecdsa_pub_key));
    assert!(result.is_err(), "expected spend to fail with wrong SLH-DSA public key hash");
}

#[test]
fn test_compile() {
    runar::compile_check(
        include_str!("SPHINCSWallet.runar.rs"),
        "SPHINCSWallet.runar.rs",
    )
    .unwrap();
}
